### PR TITLE
Add error handling documentation

### PR DESCRIPTION
**Pull request by kiootic**

### CODE REVIEW

File: docs/error-handling.md [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+`panic` should be called with a `string` or `error` argument. If the argument
+is a string, it should be in format `<module>: <message>`. For example:
+`sso: unknown provider type: example`.
+
+Panics are recovered by a middleware. Recovered panics would be logged and
+produce a 500 response if needed.
+
+### Plain error
+Plain errors are produced by error constructing functions (e.g. `errors.New`)
+or returned from dependency library. Unless otherwise required, this kind of
```
Comment (louischan-oursky) [inline docs/error-handling.md]: s/dependency/dependent

File: docs/error-handling.md [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+error should be used by default.
+
+### API error
+API errors are produced by `skyerr.Kind.NewXXX` series of functions. It should
+be used for errors responding to client.
+
+An API error consist of name, reason, message, and optionally info.
```
Comment (louischan-oursky) [inline docs/error-handling.md]: s/consist/consists

File: docs/error-handling.md [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+
+API error names are conrespond to a HTTP status code:
```
Comment (louischan-oursky) [inline docs/error-handling.md]: s/are conrespond to/correspond to

File: docs/error-handling.md [hunk 4]
```diff
@@ -0,0 +1,207 @@
+Error handling
+==============
+
+Error Kinds
+-----------
+The following kinds of error would appears in the codebase:
```
Comment (louischan-oursky) [inline docs/error-handling.md]: s/would appears/would appear

File: docs/error-handling.md [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+- Unauthorized(401): client do not have valid credentials (i.e. authentication error)
+- Forbidden(403): client's credentials are not allowed for the request (i.e. authorization error)
+- NotFound(404)
+- AlreadyExists(409)
+- TooManyRequest(429)
+- InternalError(500)
+- ServiceUnavailable(503)
+
+Reason is an identifier indicating the kind of occurred error. Each reason
+should be correspond to one error names. Reasons should be defined in the
```
Comment (louischan-oursky) [inline docs/error-handling.md]: s/one error names/one error name

File: docs/error-handling.md [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+error producing module, as opposed to centrally defined.
```
Comment (louischan-oursky) [inline docs/error-handling.md]: s/error producing module/error producing package

File: docs/error-handling.md [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+
+Message is a developer-facing message provided for ease of debugging. It should
+not contains extra useful information not found in reason/info, in order to
```
Comment (louischan-oursky) [inline docs/error-handling.md]: s/not contains/not contain

File: docs/error-handling.md [hunk 8]
```diff
@@ -0,0 +1,207 @@
...
+For some simple modules, a logic layer encapsulating component may be missing
+(e.g. user profile). In this case, the component is treated as in logic layer
+for error handling purpose.
+
+### Transport layer
+This layer handlers IO to API requests. Most components in this layer are HTTP
+handlers.
+
+Components in this layer should return API errors. It is acceptable to
+passthrough errors from data/logic layer if the error is unexpected.
```
Comment (louischan-oursky) [inline docs/error-handling.md]: s/passthrough/pass through

File: docs/error-handling.md [hunk 9]
```diff
@@ -0,0 +1,207 @@
...
+
+Errors returned by this layer would be converted to API errors.
+For non API errors, an opaque internal error(i.e. 500) would be returned to
+ensure no leakage of sensitive information.
+
+Conventions
+-----------
+
+### Error definitions
+Public errors in a module should be defined in single source file `error.go`.
```
Comment (louischan-oursky) [inline docs/error-handling.md]: s/in a module/in a package

File: docs/error-handling.md (hunk 1 above)
Comment (kiootic) [reply docs/error-handling.md]: I think dependent has incorrect meaning: 'dependent library' would mean the library that depends on it.

Comment (louischan-oursky) [reply docs/error-handling.md]: Both of them have the same meaning.

Both mean library which the application depends on.

Comment (kiootic) [reply docs/error-handling.md]: I don't think so? For example, npm use 'dependency' to indicate the package it depends on, and use 'dependent' to indicate the package that depends on it.

**Pull request merged**